spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java

```diff
@@ -94,38 +94,62 @@ public long getMessageCount() {
 	 */
 	@Override
 	public MessageGroup getMessageGroup(Object groupId) {
-		return buildMessageGroup(groupId, false);
-	}
+		MessageGroupMetadata metadata = getGroupMetadata(groupId);
+		if (metadata != null) {
 
+			MessageGroup messageGroup = getMessageGroupFactory()
+					.create(this, groupId, metadata.getTimestamp(), metadata.isComplete());
+			messageGroup.setLastModified(metadata.getLastModified());
+			messageGroup.setLastReleasedMessageSequenceNumber(metadata.getLastReleasedMessageSequenceNumber());
+			return messageGroup;
+		}
+		else {
+			return new SimpleMessageGroup(groupId);
+		}
+	}
 
-	/**
-	 * Add a Message to the group with the provided group ID.
-	 */
 	@Override
-	public MessageGroup addMessageToGroup(Object groupId, Message<?> message) {
+	public MessageGroupMetadata getGroupMetadata(Object groupId) {
 		Assert.notNull(groupId, "'groupId' must not be null");
-		Assert.notNull(message, "'message' must not be null");
+		Object mgm = this.doRetrieve(MESSAGE_GROUP_KEY_PREFIX + groupId);
+		if (mgm != null) {
+			Assert.isInstanceOf(MessageGroupMetadata.class, mgm);
+			return (MessageGroupMetadata) mgm;
+		}
+		return null;
+	}
 
-		// add message as is to the MG accessible by the caller
-		MessageGroup messageGroup = getMessageGroup(groupId);
+	@Override
+	public void addMessagesToGroup(Object groupId, Message<?>... messages) {
+		Assert.notNull(groupId, "'groupId' must not be null");
+		Assert.notNull(messages, "'messages' must not be null");
 
-		messageGroup.add(message);
+		MessageGroupMetadata metadata = getGroupMetadata(groupId);
+		SimpleMessageGroup group = null;
+		if (metadata == null) {
+			group = new SimpleMessageGroup(groupId);
```

Comment:
This case doesn't matter.
We don't remove from the group, neither return this object from the `addMessagesToGroup` method.